import "resource_common.proto";

package Heart.proto.renderpipeline;

enum InputType {
    Texture = 0;
    UniformBuffer = 1;
}

enum OutputType {
    RGBA8 = 0;
    RGBA16 = 1;
    RG16 = 2;
    RGB10A2 = 3;
    Depth = 4;
}

message PassInput {
    optional string name = 1;
    optional InputType type = 2;
}

message PassOutput {
    optional string name = 1;
    optional OutputType type = 2;
    optional uint32 index = 3;
}

message Pass {
    optional string techniqueName = 1;
    repeated PassInput inputs = 2;
    repeated PassOutput outputs = 3;
}

message PipelineBinding {
    optional uint32 fromStateID = 1;
    optional string from = 2;
    optional string to = 3;
}

message PipelineStage {
    optional string name = 1 [default = "default"];
    optional string viewName = 2 [default = "camera"];
    repeated string techniques = 3;
    repeated string outputs = 4;
	optional uint32 maxDrawCalls = 5 [default = 4000];
	optional bool clearColour = 6 [default = false];
	optional bool clearDepth = 7 [default = false];
	optional bool clearStencil = 8 [default = false];
	optional Colour clearColourValue = 9;
	optional float clearDepthValue = 10 [default = 1];
	optional uint32 clearStencilValue = 11 [default = 0];
}

message Pipeline {
    repeated PipelineStage stages = 1;
	optional uint32 totalBindings = 2 [default = 0];
}
